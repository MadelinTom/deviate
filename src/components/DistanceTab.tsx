export const DistanceTab = ({distance}: any) => {
  const {text, value} = distance;
  const kilometers = (value / 1000).toFixed(1);

  return (
    <div
      id={"distance"}
      style={{
        background: "rgba(160, 158, 151, 0.8)",
        zIndex: 1,
        display: "flex",
        borderRadius: "10px",
        alignItems: "center",
        alignSelf: "center",
        height: "50px",
        margin: "30px",
      }}
    >
      <p style={{ margin: "10px", alignSelf: "center" }}>
        {`${text} | ${kilometers} km`}
      </p>
    </div>
  );
};
