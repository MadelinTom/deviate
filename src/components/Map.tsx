import React, { useState } from "react";
import {
  GoogleMap,
  useJsApiLoader,
  DirectionsRenderer,
} from "@react-google-maps/api";

import { getDirectionsServiceOptions } from "../util";
import mapStyle from "../mapStyle.json";

const containerStyle = {
  width: "100%",
  height: "100%",
  display: "flex",
  "justify-content": "space-between",
  "flex-direction": "column-reverse",
};

const Map = () => {
  const [position, setPosition] = useState({ lat: 0, long: 0 });
  const [directionsResult, setDirectionsResult] = useState(null);
  const [runDistance, setRunDistance] = useState(null);

  const [map, setMap] = React.useState(null);
  let directionsService: any;

  const { isLoaded } = useJsApiLoader({
    id: "google-map-script",
    googleMapsApiKey:
      process.env.API_KEY || "API_KEY",
  });

  const directionsServiceCallback = (result: any, status: any) => {
    if (status == "OK") {
      console.log("dist: ", result.routes[0].legs[0].distance);
      setRunDistance(result.routes[0].legs[0].distance);
      setDirectionsResult(result);
    } else {
      console.log("directionsServiceCallback", result, status);
    }
  };

  const onLoad = React.useCallback(function callback(map) {
    const bounds = new window.google.maps.LatLngBounds();
    map.fitBounds(bounds);
    map.setOptions({ styles: mapStyle, disableDefaultUI: true, zoom: 14 });

    if ("geolocation" in navigator) {
      navigator.geolocation.getCurrentPosition(function (position) {
        console.log(position.coords.latitude);
        console.log(position.coords.longitude);
        setPosition({
          lat: position.coords.latitude,
          long: position.coords.longitude,
        });
      });
    }

    setMap(map);
  }, []);

  const onUnmount = React.useCallback(function callback(map) {
    setMap(null);
  }, []);

  const calculateNewCoords = () => {
    //@ts-ignore
    const dist = document.getElementById("distance").value * 1000;

    const directionsServiceOptions = getDirectionsServiceOptions(
      position.lat,
      position.long,
      dist || 5000
    );

    directionsService = new google.maps.DirectionsService();

    directionsService.route(
      directionsServiceOptions,
      (result: any, status: any) => directionsServiceCallback(result, status)
    );
  };

  return isLoaded ? (
    <>
      <div
        style={{
          display: "flex",
          alignSelf: "center",
          justifyContent: "center",
          height: "100%",
          width: "100%",
          zIndex: 0,
        }}
      >
        {/* For some reason this styling works */}
        <div style={{ height: "100vh", width: "100%" }}>
          <GoogleMap
            mapContainerStyle={containerStyle}
            center={new google.maps.LatLng(position.lat, position.long)}
            onLoad={onLoad}
            onUnmount={onUnmount}
            zoom={14}
          >
            {/* DISTANCE AND GENERATE BUTTON */}
            <div
              id={"contentDiv"}
              style={{
                display: "flex",
                flexDirection: "row",
                alignItems: "center",
                justifyContent: "center",
              }}
            >
              <div
                id={"form"}
                style={{
                  background: "rgba(160, 158, 151, 0.8)",
                  zIndex: 1,
                  display: "flex",
                  flexDirection: "row",
                  borderRadius: "10px",
                  alignItems: "center",
                  margin: "50px",
                }}
              >
                <form style={{ zIndex: 2, margin: "10px", minWidth: "70px" }}>
                  <label>Distance in km:</label>
                  <br />
                  <input type="number" id="distance" name="distance" />
                </form>

                <button
                  style={{
                    margin: "10px",
                    minWidth: "150px",
                    height: "50px",
                    zIndex: 2,
                  }}
                  onClick={() => calculateNewCoords()}
                >
                  Generate
                </button>
              </div>
            </div>

            {/* RUN DISTANCE */}
            {runDistance !== null ? (
              <div
                id={"distance"}
                style={{
                  background: "rgba(160, 158, 151, 0.8)",
                  zIndex: 1,
                  display: "flex",
                  borderRadius: "10px",
                  alignItems: "center",
                  alignSelf: "center",
                  height: "50px",
                  margin: "30px",
                }}
              >

                <p style={{margin: "10px", alignSelf: "center"}}>
                {/* @ts-ignore */}
                  {`${runDistance.text} | ${(runDistance.value / 1000).toFixed(1)} km`}
                </p>
              </div>
            ) : null}

            {/* DIRECTIONS */}
            {directionsResult !== null ? (
              <DirectionsRenderer
                // required
                options={{
                  directions: directionsResult || undefined,
                }}
              />
            ) : null}
          </GoogleMap>
        </div>
      </div>
    </>
  ) : (
    <></>
  );
};

export default React.memo(Map);
