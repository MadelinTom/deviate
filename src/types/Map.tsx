export interface RunDistance {
  text: string;
  value: number;
}

export interface Position {
  lat: number;
  long: number;
}