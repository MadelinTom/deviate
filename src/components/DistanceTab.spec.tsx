import { render } from "@testing-library/react";
import { RunDistance } from "../types/Map";
import { DistanceTab } from "./DistanceTab";

describe("the DistanceTab component", () => {
  const distance: RunDistance = {
    text: "test",
    value: 10,
  };

  it("should match the snapshot", () => {
    const container = render(<DistanceTab distance={distance} />);

    expect(container).toMatchSnapshot();
  });
});
