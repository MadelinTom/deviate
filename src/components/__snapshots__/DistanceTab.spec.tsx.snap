// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the DistanceTab component should match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        id="distance"
        style="background: rgba(160, 158, 151, 0.8); z-index: 1; display: flex; border-radius: 10px; align-items: center; align-self: center; height: 50px; margin: 30px;"
      >
        <p
          style="margin: 10px; align-self: center;"
        >
          test | 0.0 km
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      id="distance"
      style="background: rgba(160, 158, 151, 0.8); z-index: 1; display: flex; border-radius: 10px; align-items: center; align-self: center; height: 50px; margin: 30px;"
    >
      <p
        style="margin: 10px; align-self: center;"
      >
        test | 0.0 km
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
